src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingMiner.java

```diff
@@ -169,18 +175,38 @@ public int getDepthLimit(final Level level)
     {
         if (this.getBuildingLevel() == 1)
         {
-            return MAX_DEPTH_LEVEL_1;
+            return normalizeMaxDepth(MAX_DEPTH_LEVEL_1, level);
         }
         else if (this.getBuildingLevel() == 2)
         {
-            return MAX_DEPTH_LEVEL_2;
+            return normalizeMaxDepth(MAX_DEPTH_LEVEL_2, level);
         }
-        else if (this.getBuildingLevel() >= 3)
+        else if (this.getBuildingLevel() == 3)
+        {
+            return normalizeMaxDepth(MAX_DEPTH_MAX, level);
+        }
+        else
         {
-            return level.getMinBuildHeight() + MAX_DEPTH_LEVEL_3;
+            return normalizeMaxDepth(-100, level);
         }
+    }
 
-        return MAX_DEPTH_LEVEL_0;
+    /**
+     * Normalize the maximum depth.
+     * Make sure that the returned depth respects the world limits and follows the building setting..
+     * @param max the max depth of the given building level.
+     * @param level the world.
+     * @return the max.
+     */
+    public int normalizeMaxDepth(final int max, final Level level)
+    {
+        final int worldMaxDepth = level.getMinBuildHeight() + 5;
+        final IntSetting maxDepth = getSetting(MAX_DEPTH);
+        if (maxDepth.getValue() == maxDepth.getDefault())
+        {
+            return Math.max(worldMaxDepth, max);
+        }
```

Comment:
The check against the default value seems unnecessary.  The default is currently `-100` (which is equal to `MAX_DEPTH_MAX` anyway) which is already lower than or equal to any currently possible value of `worldMaxDepth` and `max` (until the world depth is increased again), so the line below would do the right thing even if this check were omitted.

And if in the future the world depth is increased further, then `MAX_DEPTH_MAX` would probably be decreased, as would the option default, but the current option value in existing colonies probably wouldn't (unless there was some explicit upgrade code over in the persistence, which still doesn't require anything here) and so the value would no longer equal the default and this check would be ignored anyway.